docrawl/docrawl_core.py

```diff
@@ -149,20 +135,71 @@ def _initialise_browser(self):
                 window_size_x = 1450
 
             try:
-                self.browser = webdriver.Chrome(options=self.options, service=Service(ChromeDriverManager().install()), seleniumwire_options=sw_options)
+                self.browser = webdriver.Chrome(
+                    options=self.options, service=Service(ChromeDriverManager().install()),
+                    seleniumwire_options=sw_options
+                )
             except Exception as e:
                 docrawl_logger.error(f'Error while creating Chrome instance {e}')
                 self.browser = webdriver.Chrome(options=self.options)
 
         window_size_x = 1820
-
         self.browser.set_window_size(window_size_x, 980)
 
+        browser_meta_data['browser']['pid'] = self._determine_browser_pid()
+        if browser_meta_data.get('request'):
+            browser_meta_data['request']['loaded'] = False
+        self.docrawl_client.set_browser_meta_data(browser_meta_data)
+        docrawl_logger.info(f'Browser settings: {browser_meta_data}')
+
         return self.browser
 
+    def _close_browser(self, inp):
+        """
+        Close browser (remove driver instance).
+
+        :param browser: driver instance
+        """
+        try:
+            self.browser.quit()
+        except ConnectionRefusedError:
+            pass
+        except Exception as e:
+            docrawl_logger.error(f'Error while closing the browser: {e}')
+
+        # # Remove proxy after closing browser instance
+        # proxy = {'ip': '', 'port': '', 'username': '', 'password': ''}
+        # browser_meta_data = self.docrawl_client.get_browser_meta_data()
+        # browser_meta_data['browser']['proxy'] = proxy
+        # # browser_meta_data['request']['loaded'] = False
+        # self.docrawl_client.set_browser_meta_data(browser_meta_data)
+
+    def _restart_browser(self, inp=None):
+        """Terminate any active browser (if exists/crashed) and open a new one, while retaining `browser_metadata`."""
+        if self.is_browser_active():
+            self._close_browser(inp)
+        else:
+            docrawl_logger.error("Browser crashed")
+
+        self.browser = self._initialise_browser()
+        docrawl_logger.warning("Browser restarted")
+
     def __del__(self):
         self.browser.quit()
 
+    def is_browser_active(self):
```

Comment:
Should either replace function with same name from DocrawlClient or moved there, no sense in having both.

Reply:
Agreed.